src/QirRuntime/lib/QIR/callables.cpp

```diff
@@ -0,0 +1,189 @@
+#include <assert.h>
+#include <atomic>
+#include <memory>
+#include <string.h> // for memcpy
+#include <vector>
+
+#include "__quantum__rt.hpp"
+
+/*==============================================================================
+    The types and methods, expected by QIR for tuples:
+    %TupleHeader = type { i32 }
+    declare %TupleHeader* @__quantum__rt__tuple_create(i64)
+
+    Argument passed to __quantum__rt__tuple_create is the size (in bytes) of the tuple.
+    For example:
+    ; to calculate the size of a tuple pretend having an array of them and get
+    ; offset to the first element
+    %t1 = getelementptr { %TupleHeader, %Callable*, %Array* }, { %TupleHeader, %Callable*, %Array* }* null, i32 1
+    ; convert the offset to int and call __quantum__rt__tuple_create
+    %t2 = ptrtoint { %TupleHeader, %Callable*, %Array* }* %t1 to i64
+    %0 = call %TupleHeader* @__quantum__rt__tuple_create(i62 %t2)
+
+    Notice, that the TupleHeader is placed part of the Tuple's buffer.
+==============================================================================*/
+struct QirTupleHeader
+{
+    int refCount = 0;
+};
+
+/*==============================================================================
+    Example of creating a callable
+
+    ; Single entry of a callable
+    ; (a callable might provide entries for body, controlled, adjoint and controlled-adjoint)
+    define void @UpdateAnsatz-body(
+        %TupleHeader* %capture-tuple, %TupleHeader* %arg-tuple, %TupleHeader* %result-tuple) { ... }
+
+    ; Definition of the callable
+    @UpdateAnsatz =
+        constant [4 x void (%TupleHeader*, %TupleHeader*, %TupleHeader*)*]
+        [
+            void (%TupleHeader*, %TupleHeader*, %TupleHeader*)* @UpdateAnsatz-body,
+            void (%TupleHeader*, %TupleHeader*, %TupleHeader*)* null,
+            void (%TupleHeader*, %TupleHeader*, %TupleHeader*)* null,
+            void (%TupleHeader*, %TupleHeader*, %TupleHeader*)* null
+        ]
+
+    %3 = call %Callable* @__quantum__rt__callable_create(
+        [4 x void (%TupleHeader*, %TupleHeader*, %TupleHeader*)*]* @UpdateAnsatz,
+        %TupleHeader* null)
+
+==============================================================================*/
+typedef void (*t_CallableEntry)(QirTupleHeader*, QirTupleHeader*, QirTupleHeader*);
+struct QirCallable
+{
+    static int constexpr Adjoint = 1;
+    static int constexpr Controlled = 1 << 1;

```

Comment:
These ints seem to be used to apply them to the `appliedFunctor` mask.

Maybe these pattern could be implemented in a slightly safer way by creating an enum that represents functors and inside the `ApplyFunctor` method do the logic to transform from enum value to bitmask.

The caveat is that the max enum value cannot be larger that the number of bits in an `int` but that can be ensured through an static assert.

Reply:
Another advantage of this approach would be that the following runtime assert in `ApplyFunctor` wouldn't be necessary:
`assert(functor == QirCallable::Adjoint || functor == QirCallable::Controlled)`

Reply:
Given that the values are local to the class and not exposed outside of it, I don't believe the increased safety of the two calls to ApplyFunctor outweighs the hassle of converting enums into bitfields.